opencog/atomspace/AtomSpace.h

```diff
@@ -358,6 +374,57 @@ class AtomSpace
     }
 
     /**
+     * Gets all the atoms of any type and appends them to the HandleSeq.
+     *
+     * @param appendToHandles the HandleSeq to which to append the handles.
+     *
+     * Example of call to this method:
+     * @code
+     *         std::list<Handle> atoms;
+     *         atomSpace.get_all_atoms(atoms);
+     * @endcode
+     */
+    void get_all_atoms(HandleSeq& appendToHandles) const
+    {
+        // Defer to handles by type for ATOMs and subclasses.
+        get_handles_by_type(appendToHandles, ATOM, true);
```

Comment:
These were added because there are a lot of examples of inefficient usage of the other methods, for example, ones that pass in a back_inserter rather than using the more efficient pass of the list directly which can then use the code which allocates the storage all at once.

For example:
```
atomSpace->get_handles_by_type(back_inserter(links), LINK, true);
```
would be much faster as:
```
atomSpace->get_handles_by_type(links, LINK, true);
```
because the `AtomSpace::get_handles_by_type` does a reserve on the entire size all at once before creating it's own back_inserter and passing that into `AtomTable::get_handles_by_type`. This was several times faster in my testing last year, if I recall properly. I planned on going back through the code to remove the spurious back_iterators to replace these as a separate step.

The new routines I made don't take iterators, they only take the `HandleSeq&` which means that future users won't be tempted to use the slower way out of ignorance. 

Further, the vast majority of `get_handles_by_type` code looks only for `ATOM`s, `NODE`s or `LINK`s. So by separating the common ones from more specific usages, the methods can now be optimized separately. For example, `get_all_atoms` can iterate directly from AtomTable::_atom_set and bypass the type-specific iterator used by `AtomTable::get_handles_by_type`.